### PR TITLE
Assemble direct cross-references from anchors

### PR DESCRIPTION
**Pull request by schroederc**
Each reference is resolved from an anchor's outward edges.  Grouped together by source, these are the primary data necessary for the CrossReferences API.

### CODE REVIEW

File: kythe/go/serving/pipeline/nodes/nodes.go [hunk 1]
```diff
@@ -207,3 +208,101 @@ func embedSourceKey(src *spb.VName, n *ppb.Node) *ppb.Node {
 		Edge:    n.Edge,
 	}
 }
+
+// Filter is a beam DoFn that emits *ppb.Nodes matching a set of kinds/subkinds.
+// Optionally, each processed node's facts/edges will also be filtered to
+// desired set.
+type Filter struct {
```
Comment (jrtom) [inline kythe/go/serving/pipeline/nodes/nodes.go]: I'm not sure whether there's some aspect of the existing API that requires us to have a single `ProcessElement` method to do all of these kinds of filtering, but I wonder whether the filtering code would be easier to read and understand if we broke this up into different methods: 
* one for filtering nodes by kind/subkind
* one for filtering node facts
* one for filtering (incident) edges

As it is, there's a lot of shortcut/special cases that make the code harder to read and understand.

Thoughts?

File: kythe/go/serving/pipeline/beam.go [hunk 2]
```diff
@@ -17,24 +17,161 @@
...
+			}
+			emit(ref)
+		}
+	}
+	return nil
+}
+
+func toRawAnchor(n *ppb.Node) (*srvpb.RawAnchor, error) {
+	var a srvpb.RawAnchor
+	for _, f := range n.Fact {
```
[the remark this answers is not in the document]
Comment (schroederc) [reply kythe/go/serving/pipeline/beam.go]: ✓

File: kythe/go/serving/pipeline/beam.go [hunk 3]
```diff
@@ -17,24 +17,161 @@
...
+				FilterByKind: []string{kinds.File},
+				IncludeFacts: []string{facts.Text},
+			}, k.nodes))
+
+		k.refs = beam.ParDo(s, toRefs, beam.CoGroupByKey(s, files, anchors))
+	}
+	return k.refs
+}
+
+func keyByPath(n *ppb.Node) (*spb.VName, *ppb.Node) {
```
[the remark this answers is not in the document]
Comment (schroederc) [reply kythe/go/serving/pipeline/beam.go]: It's used in later transforms.  In the cases above, both are the `*nodes.Filter`s.  The nodes are then passed into `toRefs`.  This function is used to transform a `*ppb.Node` into a `KV<*spb.VName, *ppb.Node>` for the `CoGroupByKey`.

File: kythe/go/serving/pipeline/beam.go [hunk 4]
```diff
@@ -17,24 +17,161 @@
...
+		for _, e := range n.Edge {
+			if e.GetKytheKind() == scpb.EdgeKind_CHILD_OF {
+				continue
+			}
+			ref := &ppb.Reference{
+				Source: e.Target,
+				Anchor: a,
+				Scope:  parent,
+			}
+			if k := e.GetKytheKind(); k != scpb.EdgeKind_UNKNOWN_EDGE_KIND {
```
[the remark this answers is not in the document]
Comment (schroederc) [reply kythe/go/serving/pipeline/beam.go]: ✓

File: kythe/go/serving/pipeline/beam.go [hunk 5]
```diff
@@ -17,24 +17,161 @@
...
-	return &KytheBeam{s: s, nodes: nodes.FromEntries(s, entries)}
+	return FromNodes(s, nodes.FromEntries(s, entries))
 }
 
 // Nodes returns all *ppb.Nodes from the Kythe input graph.
 func (k *KytheBeam) Nodes() beam.PCollection { return k.nodes }
+
+// References returns all derived *ppb.References from the Kythe input graph.
+func (k *KytheBeam) References() beam.PCollection {
+	if !k.refs.IsValid() {
```
[the remark this answers is not in the document]
Comment (schroederc) [reply kythe/go/serving/pipeline/beam.go]: ✓

File: kythe/go/serving/pipeline/nodes/nodes.go (hunk 1 above)
Comment (schroederc) [reply kythe/go/serving/pipeline/nodes/nodes.go]: There's no Beam requirement for this structure, but I find that it greatly simplifies the client code to have a single `Filter` over nodes.

File: kythe/go/serving/pipeline/nodes/nodes.go [hunk 6]
```diff
@@ -207,3 +208,101 @@ func embedSourceKey(src *spb.VName, n *ppb.Node) *ppb.Node {
...
+	FilterByKind, FilterBySubkind []string
+
+	IncludeFacts, IncludeEdges []string
+}
+
+// ProcessElement emit the given Node if it matches the given Filter.
```
[the remark this answers is not in the document]
Comment (schroederc) [reply kythe/go/serving/pipeline/nodes/nodes.go]: ✓

File: kythe/go/serving/pipeline/nodes/nodes.go [hunk 7]
```diff
@@ -207,3 +208,101 @@ func embedSourceKey(src *spb.VName, n *ppb.Node) *ppb.Node {
 		Edge:    n.Edge,
 	}
 }
+
+// Filter is a beam DoFn that emits *ppb.Nodes matching a set of kinds/subkinds.
+// Optionally, each processed node's facts/edges will also be filtered to
```
[the remark this answers is not in the document]
Comment (schroederc) [reply kythe/go/serving/pipeline/nodes/nodes.go]: ✓

**Comment by schroederc** [comment]
PTAL

File: kythe/go/serving/pipeline/nodes/nodes.go (hunk 1 above)
Comment (jrtom) [reply kythe/go/serving/pipeline/nodes/nodes.go]: Thanks, this is very helpful and does an excellent job of both describing the semantics and giving useful examples.

File: kythe/go/serving/pipeline/nodes/nodes.go [hunk 8]
```diff
@@ -212,10 +212,51 @@ func embedSourceKey(src *spb.VName, n *ppb.Node) *ppb.Node {
 // Filter is a beam DoFn that emits *ppb.Nodes matching a set of kinds/subkinds.
 // Optionally, each processed node's facts/edges will also be filtered to the
 // desired set.
+//
+// The semantics of the Filter are such that a "zero"-value Filter will pass all
+// Nodes through unaltered.  Part of the filter only applies if set to a non-nil
```
Comment (jrtom) [inline kythe/go/serving/pipeline/nodes/nodes.go]: s/Part/Each part

File: kythe/go/serving/pipeline/nodes/nodes.go [hunk 9]
```diff
@@ -212,10 +212,51 @@ func embedSourceKey(src *spb.VName, n *ppb.Node) *ppb.Node {
...
-
-	IncludeFacts, IncludeEdges []string
+	// FilterByKind, if non-nil, configures the filter to only pass through nodes
+	// that match one of the given kinds.
+	FilterByKind []string
+	// FilterBySubkind, if non-nil, configures the filter to only pass through
+	// nodes that match one of the given subkinds.
+	FilterBySubkind []string
+
+	// IncludeFacts, if non-nil, configures the filter remove all facts not
```
Comment (jrtom) [inline kythe/go/serving/pipeline/nodes/nodes.go]: s/remove/to remove/

File: kythe/go/serving/pipeline/nodes/nodes.go [hunk 10]
```diff
@@ -212,10 +212,51 @@ func embedSourceKey(src *spb.VName, n *ppb.Node) *ppb.Node {
...
+	// explicitly contained with the slice.
+	IncludeFacts []string
+	// IncludeEdges, if non-nil, configures the filter remove all edges with a
```
Comment (jrtom) [inline kythe/go/serving/pipeline/nodes/nodes.go]: s/remove/to remove

File: kythe/go/serving/pipeline/nodes/nodes.go (hunk 8 above)
Comment (schroederc) [reply kythe/go/serving/pipeline/nodes/nodes.go]: ✓

File: kythe/go/serving/pipeline/nodes/nodes.go (hunk 10 above)
Comment (schroederc) [reply kythe/go/serving/pipeline/nodes/nodes.go]: ✓

File: kythe/go/serving/pipeline/nodes/nodes.go (hunk 9 above)
Comment (schroederc) [reply kythe/go/serving/pipeline/nodes/nodes.go]: ✓

**Pull request merged**